### PR TITLE
Request to add DOMA object type

### PR DESCRIPTION
**Pull request by drakth**
Adding new object type DOMA.

### CODE REVIEW

**Comment by schneidermic0** [comment]
Thank you @drakth for your first proposal. I have already looked into it. I have a bunch of suggestions/questions related to naming, structuring and adding further meta information.

I have just reached out to the team responsible for DOMA to check whether the necessary information is available. I guess it make sense to check this first.

**Comment by drakth** [comment]
> Thank you @drakth for your first proposal. I have already looked into it. I have a bunch of suggestions/questions related to naming, structuring and adding further meta information.
> 
> I have just reached out to the team responsible for DOMA to check whether the necessary information is available. I guess it make sense to check this first.

Thank you!

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:25]: Aren't the fields "numbers of characters", "decimals" and "output length" numbers? If yes, I suggest to use `TYPE I`, here.

If we want to limit it to certain numbers we might want to use the annotations `$maximum` and/or `$minimum`

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:5]: Should we add an enumeration for plus and minus here?

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:44]: This seems to be the wrong type. I guess you want to use `ty_description_60` don't you.
```suggestion
      description                 TYPE zif_aff_types_v1=>ty_description_60,
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:62]:
```suggestion
      description                 TYPE zif_aff_types_v1=>ty_description_60,
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:13]: Question to DDIC experts: Does it make sense to introduce enumerations in order to get more meaningful names than jus 4 characters?

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:106]: Shall we call these `fixed_values`?
```suggestion
      "! Fixed values
      fixed_value                TYPE ty_fixed_values,
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:109]: Shall we call these `fixed_value_intervals`?
```suggestion
      "! Fixed value intervals
      fixed_value_intervals             TYPE ty_intervals_values,
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:79]: Which descriptions do you mean? Is this a different description than the one we see in the `header`?

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:76]: Use `ty_header_60_src`only for source-code based objects which require special treatment for the `abapLanguageVersion`. In other cases, just use header types without the addition `src`

```suggestion
      header                    TYPE zif_aff_types_v1=>ty_header_60,
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:85]: Just a remark, for simple types like `abap_bool` we haven't introduce a dedicated type, yet. At least, if we didn't have any reuse.

**Review by schneidermic0 — CHANGES_REQUESTED** [review]
Thank you for your PR. 
I think that's a good way to start. :)

I have reached out to developers responsible for domains. They provided me their first feedback:

- There are some formatting missing (e.g., output style for decimal values or formats for time)
- The proposal does not consider domain appends, yet. I suggest we start with plain DOMAs first and add information relevant for domain appends in a second iteration.

I have also added also some comments more general to the file formats. See below. I hope the comments help.

Let us know if you face questions.

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:9]: Sounds good to me, I think the abap_bool is missing from the modification?

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:25]: From what I see in table D001L and in transaction SE11 they seem to be NUMC, however if you think TYPE I would be better its ok by me.

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:44]: Yes, you are correct.

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:106]: Sounds good to me!

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:109]: Also its ok by me.

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:79]: Yes, you are right, description needs to be removed I think. I could remove descriptions and rename header to description, what do you think?

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [reply file-formats/doma/type/zif_aff_doma_v1.abap]: Oh, sorry, this was my fault. I made a wrong suggestion. I would use `fixed_values`. I will add a new suggestion in your latest version.

**Review by schneidermic0 — CHANGES_REQUESTED** [review]
Cool, thanks for adapting your proposal. 

For the `fixed_values` I missed to add the `s` in my initial proposal.

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [reply file-formats/doma/type/zif_aff_doma_v1.abap:79]: I would just keep the `header` structure. There is a field `description` within the header. It actually shall represent the description of the domain

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [reply file-formats/doma/type/zif_aff_doma_v1.abap:25]: For other object types we suggested to use integer with `$maximum` and `$minimum`. We should use the numeric only, if we want to keep the leading zeros.

Like this:

 ```suggestion
  "! $minimum 0
  "! $maximum 999999
  TYPES ty_numb_char            TYPE i.

  "! <p class="shorttext">Decimals</p>
  "! Decimals
  "! $minimum 0
  "! $maximum 999999
  TYPES ty_decimals             TYPE i.

  "! <p class="shorttext">Output Length</p>
  "! Output Length
  "! $minimum 0
  "! $maximum 999999
  TYPES ty_output_length        TYPE i.
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [reply file-formats/doma/type/zif_aff_doma_v1.abap:9]: > I think the abap_bool is missing from the modification?

I am not sure. What is your question?

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (huber-nicolas) [inline file-formats/doma/type/zif_aff_doma_v1.abap:8]: typo: case sensi**_t_**ive

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (huber-nicolas) [inline file-formats/doma/type/zif_aff_doma_v1.abap:102]: Typo: value table

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:9]: Do you mean to change it to:
`TYPES ty_case_sensitive TYPE abap_bool` ?

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:8]: Thank you!

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap]: Thank you!

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:5]: Would be something like the example below be ok?
`  "! <p class="shorttext">Sign</p>
  "! Sign
  "! $values {@link zif_aff_doma_v1.data:co_sign}
  TYPES ty_sign TYPE c LENGTH 1.

  CONSTANTS:
    "! <p class="shorttext">Sign</p>
    "! Sign
    BEGIN OF co_sign,
      "! <p class="shorttext">Plus</p>
      "! Plus
      Plus TYPE ty_sign VALUE '+',
      "! <p class="shorttext">Minus</p>
      "! Minus
      Minus TYPE ty_sign VALUE '-',
    END OF co_sign.`

Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:5]: Would something like the example above be ok?

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap]: Sounds good to me!

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (albertmink) [inline file-formats/doma/type/zif_aff_doma_v1.abap:57]:
```suggestion
  TYPES:
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (albertmink) [inline file-formats/doma/type/zif_aff_doma_v1.abap:43]:
```suggestion
  TYPES:
```

**Review by albertmink — COMMENTED** [review]
I found some trailing whitespaces.

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (albertmink) [inline file-formats/doma/type/zif_aff_doma_v1.abap:112]:
```suggestion
      fixed_value_intervals             TYPE ty_intervals_values,
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (albertmink) [inline file-formats/doma/type/zif_aff_doma_v1.abap:109]:
```suggestion
      fixed_values                TYPE ty_single_values,
```

**Review by albertmink — COMMENTED** [review]
I couldn't activate the INTF. So maybe there is a type miss match?

**Comment by albertmink** [comment]
I just generated a JSON schema from the ABAP type and sampled JSON data [here](https://gist.github.com/albertmink/ae8853700c011fd3d9af94bf96f3e470)
Hopefully, this helps to get a better feeling of the actual JSON data that is specified in this PR.

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [reply file-formats/doma/type/zif_aff_doma_v1.abap:9]: Yes. Oh, I see, now, my initial proposal did not change anything. Here is what I meant:

```suggestion
  TYPES ty_case_sensitive       TYPE abap_bool.
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [reply file-formats/doma/type/zif_aff_doma_v1.abap:5]: Yes, in pricinple, I think this is the way to go. Usually, we take the values which are stored on the database (as far as I can see this would be `x` and `<space>` in this case. Therefore, I think I would end up with a proposal like this:

```
  CONSTANTS:
    "! <p class="shorttext">Sign</p>
    "! Sign
    BEGIN OF co_sign,
      "! <p class="shorttext">Plus</p>
      "! Plus
      Plus TYPE ty_sign VALUE ' ',
      "! <p class="shorttext">Minus</p>
      "! Minus
      minus TYPE ty_sign VALUE 'X',
    END OF co_sign.
```

Another option would be that we used just a boolean to specify whether the field allows a sign (i.e., negative values) or not. In this case we wouldn't need constants. What do you think?

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:109]: Yes, I think you are right.

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:5]: While the boolean option is indeed shorter, by quickly looking at the code, I think the constant one is easier to understand.

Comment (schneidermic0) [reply file-formats/doma/type/zif_aff_doma_v1.abap:5]: If we move it to a boolean, we most probably will have also to rename the field to something like `negative_values_supported`(or similar). 

However, I'll reach out to some DDIC experts and ask whether they have an opinion on this. It might be that it take some time until I get a response due to the Holiday season.

Comment (schneidermic0) [reply file-formats/doma/type/zif_aff_doma_v1.abap:5]: I have got already feedback by the DDIC experts. They would prefer the boolean option so that we can see in the JSON following information:

```
  „supportsNegativeValues“: true
```

What do you think? Is this fine for you?

Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:5]:
```suggestion
  TYPES ty_negativeValues TYPE abap_bool.
```

Its fine by me, would this be ok?

Comment (schneidermic0) [reply file-formats/doma/type/zif_aff_doma_v1.abap:5]: Almost. Just an underscore is missing :)
```suggestion
  TYPES ty_negative_values TYPE abap_bool.
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:85]: We have to change is here as well

```suggestion
      negative_values                      TYPE ty_negative_values,
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:103]:
```suggestion
      conversion_routine                 TYPE ty_conv_routine,
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:94]: I am wondering. Didn't we change this already?

```suggestion
      length                 TYPE ty_length,
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:5]: thanks!

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:94]: Not that I remember, and I couldn't find anything in the history, unless I missed it :P

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [inline file-formats/doma/type/zif_aff_doma_v1.abap:19]: Since we changed the field `numb_char` to `length`, I suggest to also rename the type to `ty_length`

```suggestion
 TYPES ty_numb_char            TYPE i.
```

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (schneidermic0) [reply file-formats/doma/type/zif_aff_doma_v1.abap]: Maybe, I was wrong. Thanks anyhow :)

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap:19]: I'm sorry I'm kind of lost, you mean

```suggestion
 TYPES ty_numb_char            TYPE ty_length.
```
?

Comment (schneidermic0) [reply file-formats/doma/type/zif_aff_doma_v1.abap:19]: Sorry, my fault. I didn't adapt my suggestion. This is what I meant 
```suggestion
 TYPES ty_length            TYPE i.
```

Reason: We renamed the type for field `length` to `ty_length` but didn't adapt the type.

P.S.: I remove my previous suggestion, because it does not make any sense :(

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap]: Forgot to comment, this should be fixed.

**Review by schneidermic0 — CHANGES_REQUESTED** [review]
In general, it looks good to me, now. Thanks :) I also asked my colleagues to have another look at it.

We still have to fix one syntax error. See my comment below.

**Comment by schneidermic0** [comment]
If you like, you can also add an example object as JSON representation to this pull request. 

As soon as we have merged the changes of this PR. We (SAP) generate a corresponding JSON-schema file and add it to the repository.

File: file-formats/doma/type/zif_aff_doma_v1.abap
Comment (drakth) [reply file-formats/doma/type/zif_aff_doma_v1.abap]: done! :)

**Comment by drakth** [comment]
> If you like, you can also add an example object as JSON representation to this pull request.
> 
> As soon as we have merged the changes of this PR. We (SAP) generate a corresponding JSON-schema file and add it to the repository.

I've added the example JSON file, hopefully without errors.

File: file-formats/doma/examples/z_aff_example_doma.doma.json
Comment (albertmink) [inline file-formats/doma/examples/z_aff_example_doma.doma.json:24]: You might copy&paste the correct formatting from here
```
{
  "formatVersion": "1",
  "header": {
    "description": "Example DOMA for ABAP file formats",
    "originalLanguage": "en"
  },
  "negative_values": false,
  "case_sensitive": false,
  "data_type": "CHAR",
  "length": 3,
  "decimals": 2,
  "output_length": 3,
  "conversion_routine": "",
  "value_table": "",
  "fixed_values": {
    "fixed_value": "D",
    "description": "Test fixed values description"
  },
  "fixed_value_intervals": {
    "low_limit": "1",
    "high_limit": "5",
    "description": "Test interval description"
  }
}
```

Comment (albertmink) [reply file-formats/doma/examples/z_aff_example_doma.doma.json:24]: with an empty line at the end 😄

File: file-formats/doma/examples/z_aff_example_doma.doma.json
Comment (schneidermic0) [inline file-formats/doma/examples/z_aff_example_doma.doma.json:14]: Just a minor. Since these fields are empty, the usually won't be serialized.
```suggestion
```

**Review by schneidermic0 — COMMENTED** [review]
I found just one more minor issue in the new example.

File: file-formats/doma/examples/z_aff_example_doma.doma.json
Comment (drakth) [reply file-formats/doma/examples/z_aff_example_doma.doma.json]: Should be fixed now. Thank you.

File: file-formats/doma/examples/z_aff_example_doma.doma.json
Comment (drakth) [reply file-formats/doma/examples/z_aff_example_doma.doma.json]: Fixed!

**Review by schneidermic0 — APPROVED** [review]
Thanks, @drakth for your contribution. 

Looks good to me. 

I am still waiting for feedback by my colleagues. But I guess we can push this as first version. I expect some (small) changes in the future like reordering of fields or maybe grouping them. But I think it is a very good starting point. Thank you. :)

**Comment by drakth** [comment]
> I have got the OK from our DDIC experts, too.
> 
> @drakth Welcome as contributor! :) Thank you.

Thank you very much!

**Pull request merged**
